File: SetMatrixZeroes.py
class Solution:
    def setZeroes(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: void Do not return anything, modify matrix in-place instead.
        """
        m = len(matrix)
        n = len(matrix[0])
        col0 = False
        for i in range(m):
            if matrix[i][0]==0:
                col0 = True
            for j in range(1, n):
                if matrix[i][j] == 0:
                    matrix[i][0] = 0
                    matrix[0][j] = 0
        for i in range(m-1, -1 ,-1):
            for j in range(n-1, 0, -1):
                if matrix[i][0]==0 or matrix[0][j]==0:
                    matrix[i][j]=0
            if col0:
                matrix[i][0] = 0
                
if __name__=='__main__':
    m = int(input())
    matrix = []
    for i in range(m):
        raw = list(input())
        matrix.append(raw)
    print(matrix)
    solution = Solution()
    solution.setZeroes(matrix)
    print(matrix)

File: SetMatrixZeroes.cpp
/*
 * O(m+n)
*/
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    void setZeroes(vector<vector<int> >& matrix) {
        
        int m=matrix.size(), n=matrix[0].size();
        
        int row[m], clumn[n];
        for(int i=0;i<m;i++) row[i]=0;
        for(int i=0;i<n;i++) clumn[i]=0;
        
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(matrix[i][j]==0){
                    row[i]++;
                    clumn[j]++;
                }
            }
        }
        
        for(int i=0;i<m;i++){
            if(row[i]){
                for(int j=0;j<n;j++){
                    matrix[i][j]=0;
                }
            }
        }
        
        for(int j=0;j<n;j++){
            if(clumn[j]){
                for(int i=0;i<m;i++){
                    matrix[i][j]=0;
                }
            }
        }
    }
};

int main(){
    int m, n;
    cin>>m>>n;
    
    vector<vector<int> > matrix;
    for(int i=0;i<m;i++){
        vector<int> row;
        for(int j=0;j<n;j++){
            int num;
            cin>>num;
            row.push_back(num);
        }
        matrix.push_back(row);
    }
    
    Solution *solution=new Solution();
    solution->setZeroes(matrix);
    for(int i=0;i<m;i++){
        for(int j=0;j<n;j++){
            cout<<matrix[i][j]<<" ";
        }
        cout<<endl;
    }
}
